Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -774,20 +774,124 @@ namespace OpenInfraPlatform {
+					const carve::math::Matrix& pos,
+					std::shared_ptr<carve::input::PolyhedronData>& polygon, //Carve polygon of the converted face
+					std::map<std::string, uint32_t> polygonIndices // Contains polygon indices of vertices (x,y,z converted to string)
+				)  const noexcept(false)
+				{
+					if (advancedFace.expired()) {
+						throw oip::ReferenceExpiredException(advancedFace);
+					}
+
+					// get reference to IfcFaceBound's (attribute 1)
+					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds = advancedFace->Bounds;
+					// if one face bound is not of the allowed type: throw; else: continue and use already converted faceBoundLoops
+					if (!std::all_of(ifcFaceBounds.begin(), ifcFaceBounds.end(), [](const auto& ifcFaceBound) {return
+						ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcEdgeLoop>() || ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcVertexLoop>(); }))
+					{
+						throw oip::InconsistentModellingException(advancedFace, "IfcAdvancedFace has a loop type as face boundary which is not allowed.");
+					}
+
+					// get attribute 3: indicates whether the sense of the surface normal agrees with the sense of the topological normal (face bound)
+					// ToDo: take sameSense into account by the addFace-construction
+					const bool sameSense = advancedFace->SameSense;
+
+					// get attribute 2
+					const EXPRESSReference<typename IfcEntityTypesT::IfcSurface>& surface = advancedFace->FaceSurface;
 
-						addArbitraryFaceToPolyhedronData(face, faceBoundLoops, polygon, polygonIndices);
+					// the surface (geometry) is trimmed by the faceBoundLoops (topology, represented in geometrical description)
+					// Loop geometry should be consistent with the face geometry (buildingSMART), thus the loop vertices should be part of the surface.
+					// 7 surface types are valid; each type does probably need its individual handling (how the surface is trimmed by the boundary)
+					if (surface.isOfType<typename IfcEntityTypesT::IfcPlane>())
+					{
+						// loop points lie in a plane (the IfcPlane), thus the loop points can be triangulated
+						addArbitraryFaceToPolyhedronData(advancedFace, faceBoundLoops, polygon, polygonIndices);
+					}
+					else if (surface.isOfType<typename IfcEntityTypesT::IfcBSplineSurface>())
+					{
+						// get surface geometry into inputDataFaceSurface
+						std::shared_ptr<ItemData> inputDataFaceSurface(new ItemData);
+						convertIfcBSplineSurface(surface.as<typename IfcEntityTypesT::IfcBSplineSurface>(), pos, inputDataFaceSurface);
+
+						// ASSUMPTION: the B-Spline-Surface boundary is already coincident with the loop boundary
+						// ToDo: in general case, the B-Spline-Surface has to be trimmed by the loop boundary (or at least checked to be coincident)
+
+						// append surface-faces to target polygon
+						mergePolyhedronsIntoOnePolyhedron(inputDataFaceSurface, polygon);

```

Comment:
Please make this a member function of `ItemData`.

Reply:
Done in commit 237f2b3.